source/custom/nidaqmx_service.custom.cpp

```diff
@@ -1,4 +1,106 @@
 #include <nidaqmx/nidaqmx_service.h>
 
 namespace nidaqmx_grpc {
+//---------------------------------------------------------------------
+//---------------------------------------------------------------------
+::grpc::Status NiDAQmxService::ReadAnalogF64(::grpc::ServerContext* context, const ReadAnalogF64Request* request, ReadAnalogF64Response* response)
+{
+  if (context->IsCancelled()) {
+    return ::grpc::Status::CANCELLED;
+  }
+  try {
+    auto task_grpc_session = request->task();
+    TaskHandle task = session_repository_->access_session(task_grpc_session.id(), task_grpc_session.name());
+    int32 num_samps_per_chan = request->num_samps_per_chan();
+    float64 timeout = request->timeout();
+    int32 fill_mode;
+    switch (request->fill_mode_enum_case()) {
+      case nidaqmx_grpc::ReadAnalogF64Request::FillModeEnumCase::kFillMode: {
+        fill_mode = static_cast<int32>(request->fill_mode());
+        break;
+      }
+      case nidaqmx_grpc::ReadAnalogF64Request::FillModeEnumCase::kFillModeRaw: {
+        fill_mode = static_cast<int32>(request->fill_mode_raw());
+        break;
+      }
+      case nidaqmx_grpc::ReadAnalogF64Request::FillModeEnumCase::FILL_MODE_ENUM_NOT_SET: {
+        return ::grpc::Status(::grpc::INVALID_ARGUMENT, "The value for fill_mode was not specified or out of range");
+        break;
+      }
+    }
+
+    uInt32 array_size_in_samps = request->array_size_in_samps();
+    if (array_size_in_samps == 0) {
```

Comment:
The custom code starts here.

Reply:
This seems pretty reasonable to me. Implementing the codegen might be a challenge, especially if the different flavors of read are different. And also that we'd need a fair amount of testing to cover the different codepaths. This is probably "nice to have" and "something we could add later", but maybe we can get it in.

If the implementations for different read methods *are* the same, is it possible we could pull this out into helper methods? So the generated code could just be:

```c++
auto num_samps_per_chan = calculate_num_samps_per_chan(library, task, request->num_samps_per_chan());
auto array_size_in_samps = calculate_array_size_in_samps(library, task, num_samps_per_chan, request->array_size_in_samps());
```
I guess we'd have to use exceptions to handle errors, but that's probably doable. A FailedStatusException could be useful in other places.

In that case we *could* use something like the `hardcoded` or `custom-code` mechanisms to just code the one-liner in the metadata. But it would also be fine to just have a special `daq-read-samps-per-chan` and `daq-array-size-in-samps` mechanisms in the metadata/mako but have the implementation be very simple.

Also, this is sort of 2 separate features right? We could minimally just have something like:

```c++
auto array_size_in_samps = request->array_size_in_samps() 
    ? request->array_size_in_samps() 
    : (num_samps_per_chan * get_number_of_channels(library, task);
```